packages/frontend/src/contexts/AppContext/useTxHistory.ts

```diff
@@ -25,86 +27,200 @@ const cacheKey = 'recentTransactions:v000'
-    return JSON.parse(value).map((obj: Transaction) => Transaction.fromObject(obj))
+    if (!value) return undefined // Handle empty string if needed
+    return JSON.parse(value).map((obj: any) => Transaction.fromObject(obj)) as Transaction[]
   },
 }
 
 const useTxHistory = (defaultTxs: Transaction[] = []): TxHistory => {
-  const [transactions, setTransactions, clear] = useLocalStorage<Transaction[]>(
-    cacheKey,
-    defaultTxs,
-    localStorageSerializationOptions
-  )
+  const [transactions, setTransactions] = useState<Transaction[] | undefined>(defaultTxs)
+
+  useEffect(() => {
+    // on mount, load from local storage
+    const storedTxs = localStorage.getItem(cacheKey)
+    if (storedTxs) {
+      setTransactions(localStorageSerializationOptions.deserializer(storedTxs))
+    }
+  }, [])
+
+  useEffect(() => {
+    // on transactions change, save to local storage
+    if (transactions) {
+      localStorage.setItem(cacheKey, localStorageSerializationOptions.serializer(transactions))
+    }
+  }, [transactions])
+
+  function clear() {
+    try {
+      localStorage.removeItem(cacheKey)
+    } catch (error) {
+      console.error(error)
+    }
+  }
 
   function filterSortAndSetTransactions(tx: Transaction, txs?: Transaction[], hashFilter?: string) {
-    const filtered = filterByHash(txs, hashFilter)
-    setTransactions(sortByRecentTimestamp([...filtered, tx]).slice(0, 3))
+    setTransactions(prevTransactions => {
+      const currentTxs = txs ?? prevTransactions ?? []
+      const filtered = filterByHash(currentTxs, hashFilter)
+      return sortByRecentTimestamp([...filtered, tx]).slice(0, 4)
+    })
   }
 
-  const addTransaction = useCallback(
-    (tx: Transaction) => {
-      // disable this feature
-      return
-      // If tx exists with hash == tx.replaced, remove it
-      const match = find(transactions, ['hash', tx.replaced])
-      filterSortAndSetTransactions(tx, transactions, match?.hash)
-    },
-    [transactions]
-  )
-
-  const removeTransaction = useCallback(
-    (tx: Transaction) => {
-      // If tx exists with hash == tx.replaced, remove it
-      const filtered = filterByHash(transactions, tx.hash)
-      setTransactions(sortByRecentTimestamp(filtered).slice(0, 3))
-    },
-    [transactions]
-  )
-
-  const updateTransaction = useCallback(
-    (tx: Transaction, updateOpts: UpdateTransactionOptions, matchingHash?: string) => {
-      // disable this feature
-      return
-      for (const key in updateOpts) {
-        tx[key] = updateOpts[key]
+  function addTransaction(tx: Transaction) {
+    const match = find(transactions, ['hash', tx.replaced])
+    filterSortAndSetTransactions(tx, transactions, match?.hash)
+  }
+
+  function removeTransaction(tx: Transaction) {
+    setTransactions(prevTransactions => {
+      if (!prevTransactions) return []
+      const filtered = filterByHash(prevTransactions, tx.hash)
+      return sortByRecentTimestamp(filtered).slice(0, 4)
```

Comment:
suggestion: store the max number as a variable since it's used in multiple places

```suggestion
return sortByRecentTimestamp(filtered).slice(0, maxLimit)
```